File: leetcode/621.task-scheduler.py
# @leet start
class Solution:
    def leastInterval(self, tasks: list[str], n: int) -> int:
        CODE_POINT_A = ord("A")

        max_count = 0
        num_with_max_count = 0
        counts = [0] * 26

        for task in tasks:
            idx = ord(task) - CODE_POINT_A
            counts[idx] += 1

            if counts[idx] == max_count:
                num_with_max_count += 1
            elif counts[idx] > max_count:
                max_count += 1
                num_with_max_count = 1

        return max(len(tasks), num_with_max_count - 1 + max_count + n * (max_count - 1))


# @leet end

File: leetcode/621.task-scheduler.js
// @leet start
/**
 * @param {character[]} tasks
 * @param {number} n
 * @return {number}
 */
var leastInterval = function (tasks, n) {
  const counts = new Map();
  let maxCount = 0;
  let numWithMax = 0;

  for (const task of tasks) {
    const newCount = 1 + (counts.get(task) ?? 0);
    counts.set(task, newCount);

    if (newCount === maxCount) {
      numWithMax++;
    } else if (newCount > maxCount) {
      maxCount = newCount;
      numWithMax = 1;
    }
  }

  return Math.max((n + 1) * (maxCount - 1) + numWithMax, tasks.length);
};
// @leet end
